### PR TITLE
[MPS] Add upsample_bicubic2d as Metal op

### PR DESCRIPTION
**Pull request by malfet**
More or less literal copy-n-paste of https://github.com/pytorch/pytorch/blob/c33b0580e6a702be0cd5be691b3b465da012aa34/aten/src/ATen/native/cuda/UpSampleBicubic2d.cu#L24
and
https://github.com/pytorch/pytorch/blob/c33b0580e6a702be0cd5be691b3b465da012aa34/aten/src/ATen/native/cuda/UpSampleBicubic2d.cu#L99
Missing `uint8` implementation mimics CUDA behavior
Initial version coded live in https://www.youtube.com/watch?v=shi6Kb5xxvk
Later refinements:
 - Switch from 2D dispatch to 1D one (to match CUDA behavior)
 - Added batch + channel loops
 - Fixed scale computation to match align corners behavior
 - Added backward implementation

Backward implementation again, mimics CUDA, so it has issues precision issue for `torch.half` as well as a somewhat slow simulation of atomic adds using atomic compare and exchange of the pair of adjacent values, i.e.
```metal
emplate <typename T>
static inline void atomic_add_helper(
    device atomic<int>* data,
    long offset,
    float value) {
  auto ptr = data + (offset >> 1);
  auto old = atomic_load_explicit(ptr, memory_order_relaxed);
  union {
    int i;
    T t[2];
  } val;
  do {
    val.i = old;
    val.t[offset & 1] += static_cast<T>(value);
  } while (!atomic_compare_exchange_weak_explicit(
      ptr, &old, val.i, memory_order_relaxed, memory_order_relaxed));
}
```
Bump basic Metal language version to 3.0, as it's supported on MacOS13 and that's the first version that has `atomic_float`

### CODE REVIEW

**Comment by malfet** [comment]
It now has an unexpected successes, which is a good sign, next step is to fix it for uint8 dtype.

File: aten/src/ATen/native/mps/operations/UpSample.mm [hunk 1]
```diff
@@ -216,6 +217,232 @@ static void upsample_out_template(const Tensor& input,
...
+  }
+}
+
+static void upsample_bicubic_out_template(const Tensor& input,
+                                          IntArrayRef output_size,
+                                          bool align_corners,
+                                          std::optional<double> scale_h_opt,
+                                          std::optional<double> scale_w_opt,
+                                          const Tensor& output) {
+  if (input.numel() == 0) {
```
Comment (albanD) [inline aten/src/ATen/native/mps/operations/UpSample.mm:418]: Do you need to check both input and output numel here?

**Review by albanD — APPROVED** [review]
Integration sounds good, only small nit on edge case.
I didn't double check the convolution algorithm but I guess that CI is enough to validate it?

**Pull request closed without merging**